spark/v3.2/spark/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryPartitionPathsIndex.scala

```diff
@@ -0,0 +1,122 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+import org.apache.hadoop.fs._
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.types.StructType
+import scala.collection.mutable
+
+
+/**
+ * Implement our own in-memory index which will only list directories to avoid unnecessary file
+ * listings. Should ONLY be used to get partition directory paths. Uses table's schema to only
+ * visit partition dirs using number of partition columns depth recursively. Does NOT return files
+ * within leaf dir.
+ */
+class InMemoryPartitionPathsIndex(
+                         sparkSession: SparkSession,
+                         rootPath :Path,
+                         parameters: Map[String, String],
+                         userSpecifiedSchema: StructType,
+                         fileStatusCache: FileStatusCache = NoopCache,
+                         override val metadataOpsTimeNs: Option[Long] = None)
+  extends PartitioningAwareFileIndex(
+    sparkSession, parameters, Some(userSpecifiedSchema), fileStatusCache) {
+
+  override val rootPaths = Seq(rootPath)
+
+  @volatile private var cachedLeafFiles: mutable.LinkedHashMap[Path, FileStatus] = _
+  @volatile private var cachedLeafDirToChildrenFiles: Map[Path, Array[FileStatus]] = _
+  @volatile private var cachedPartitionSpec: PartitionSpec = _
+
+  refresh0()
+
+  override def partitionSpec(): PartitionSpec = {
+    if (cachedPartitionSpec == null) {
+        cachedPartitionSpec = inferPartitioning()
+    }
+    logTrace(s"Partition spec: $cachedPartitionSpec")
+    cachedPartitionSpec
+  }
+
+  override def allFiles(): Seq[FileStatus] = cachedLeafFiles.values.toSeq
+
+  override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = cachedLeafFiles
+
+  override protected def leafDirToChildrenFiles: Map[Path, Array[FileStatus]] = cachedLeafDirToChildrenFiles
+
+  override def refresh(): Unit = {
+    fileStatusCache.invalidateAll()
+    refresh0()
+  }
+
+  private def refresh0(): Unit = {
+    val files = listLeafDirs(rootPath, 0)
+    cachedLeafFiles =
+      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
+    cachedLeafDirToChildrenFiles = files.map( f => (f.getPath, Array(createEmptyChildDataFileStatus(f))))
+      .toMap
+    cachedPartitionSpec = null
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case hdfs: InMemoryFileIndex => rootPaths.toSet == hdfs.rootPaths.toSet
+    case _ => false
+  }
+
+  override def hashCode(): Int = rootPaths.toSet.hashCode()
+
+  /**
+   * recursively lists only the partition dirs. Uses the number of partition cols
```

Comment:
If I understand it correctly, the original method here is parallel, now this changes to single process.  Is there any way to use more of the existing code, ie by passing a filter to the original bulkList methods?

Reply:
@szehon-ho 
correct, original method may distribute the task to list files.

However, in our implementation as we **don't** list files and _only_ list directories, we don't need parallelism as listing  just dirs will be quicker or not that many compared to listing files in each dir.

Reply:
@szehon-ho 
Regarding using existing code, I tried that and unfortunately InMemoryFileIndex constructor called refresh -> listFiles,
due to current implementation it doesn't allow overriding them.

Reply:
Thanks for benchmarking.  

I'm still not entirely sure why old code is 12 min vs old code being < 1 second?   My thought is, the old code spawned a spark job?  The InMemoryFileIndex list seems to also call HadoopFsUtils.parallelListLeafFiles, which seems to do a similar thing, as long as threshold was not set.

yea about code re-use, I'm not entirely sure all the Hadoop-side FS stuff.  While this may work on S3, there looks like some special handling for special Hadoop FileSystem inside HadoopFsUtils.parallelListLeFiles, which I fear we may not handle.  Hence suggestion to call existing code HadoopFsUtils.parallelListLeafFiles

Reply:
@szehon-ho the old code does spawn a spark job based on the parallelism threshold, which was not met in my benchmark. Parallelism makes sense for listLeafFiles as it gets distributed to executors for each folder/partition to list the files. But in our case, as we don't _listFiles_ but only dirs, I don't think we need paralellism or spark job just to list dirs

Reply:
regarding `HadoopFsUtils.parallelListLeafFiles`, we cannot use that with filter because the filter is applied _after_ leaf files are listed, we are trying to avoid that in this PR.

Reply:
https://github.com/apache/spark/blob/46332d985e52f76887c139f67d1471b82e85d5ca/core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala#L174

HadoopFsUtils.parallelListLeafFiles has special code only if the file system is distributed/view FS, which in our case of Add_files will not be the case, I think we should be file doing `fs.listStatus` on Hadoop fs and should handle for non s3. thoughts?

Reply:
OK makes sense.  I was looking the other day and we have custom code in DeleteFileSparkAction.listFilesRecursively.  I'm a bit behind and will need to take another look later this week, unless someone else here has more time first